### PR TITLE
Allow write_file to sanitize spaces and newlines

### PR DESCRIPTION
**Pull request by ssbarnea**
Needed-By: https://github.com/ansible-community/antsibull-docs/issues/305

### CODE REVIEW

**Review by felixfontein — CHANGES_REQUESTED** [review]
I think this is the wrong place for such functionality. This should be done by the caller before calling `write_file()`.

Also the implementation is very inefficient. Check for example with `" "*100000 + "x" + " "*100000`. Compare for example to `content = '\n'.join([l.rstrip(' \t') for l in content.splitlines()])`, which is *so* much faster for this example.

File: src/antsibull_core/utils/io.py [hunk 1]
```diff
@@ -71,10 +72,17 @@ async def copy_file(
     flog.debug("Leave")
 
 
-async def write_file(filename: StrOrBytesPath, content: str) -> None:
+async def write_file(
+    filename: StrOrBytesPath, content: str, /, sanitize: bool = False
```
Comment (gotmax23) [inline src/antsibull_core/utils/io.py:76]: Technically speaking, making `filename` and `content` positional-only arguments is a breaking change.

File: src/antsibull_core/utils/io.py [hunk 2]
```diff
@@ -71,10 +72,17 @@ async def copy_file(
...
+) -> None:
+    """Dumps a generated text file to disk, allows sanitization on trimming spaces."""
     flog = mlog.fields(func="write_file")
     flog.debug("Enter")
 
+    if sanitize:
+        content = re.sub(r"[ \t]+$", "", content, flags=re.MULTILINE)
+        content = content.rstrip("\n\r") + "\n"
```
Comment (gotmax23) [inline src/antsibull_core/utils/io.py:84]: I agree with Felix that perhaps this isn't the best place to do this, but maybe a separate `sanitize_trailing_spaces()` function in antsibull_core would make sense

File: changelogs/fragments/165-write-file-sanitize.yml [hunk 3]
```diff
@@ -0,0 +1,2 @@
+minor_changes:
+  - Allow write_file to sanitize spaces and newlines. (https://github.com/ansible-community/antsibull-docs/issues/305)
```
Comment (gotmax23) [inline changelogs/fragments/165-write-file-sanitize.yml:2]:
```suggestion
  - Allow write_file to sanitize spaces and newlines (https://github.com/ansible-community/antsibull-docs/issues/305).
```

is the format we generally follow.

**Comment by gotmax23** [comment]
Is this still relevant? I think @ssbarnea's original issue was solved differently in antsibull-docs?

**Comment by gotmax23** [comment]
Either way, thank you for the contribution.

**Comment by felixfontein** [comment]
Closing since this is apparently no longer needed. Thanks for your contribution!

**Pull request closed without merging**